### ISSUE TITLE
Documentation review

### ISSUE THREAD

**Issue by cwapshere** [Microsoft/MIMWAL#1]
Hi all. Jef asked if I could do some doco review - very happy to help. I've only got through the Activities section so far but wanted to post these comments - hopefully this is an ok place to do it.

**Build and Deployment**

*Note: If you get any access denied errors from FIM Service during the creation of AIC's, you may have forgotten to correct the typo "IsAuthoriztionActivity" to "IsAuthorizationActivity" in the built-in MPR called "Administration: Administrators control configuration related resources".*

I didn’t know myself that such a typo existed and I just create my own AIC policy. Is there a link to a technote about this? Rather than “You may have forgotten” instead link to the technote. (I’m surprised a patch would not have fixed this.)

**Create Resource Activity**

*Iteration
Optional. This is a lookup or function expression returning a collection of values to iterate over. use of iteration disables publication of created resource Id ("Target for Created Resource ID") and conflicting resource Id ("Target for Conflicting Resource ID").*

In the “Delete Resource” activity there is a link to the Iteration page from this sub-section.

*While the activity supports iteration, it is best to refrain from creating more than one resources in a single activity.*

Should be “resource” not “resources”.

Why is it best? I only ever used this activity to create single resources myself, but the functionality must be there for some reason. Would be good to add a bit more explanation about why it is recommended – the main things I can think of right now is it’s much easier to track what’s going on, and there is no chance of accidentally spawning creation of thousands of objects. Or alternatively add a comment about whatever specific use case this feature was added for and state you wouldn’t use it otherwise.

BTW this group creation example is a really good one – I’ve recently been finding out just how difficult(/impossible) something like this is for certain competitor products.

**Generate Unique Value Activity**

Noticed one instance of “FIMServive” on the page.

It’s good you talk about bulk updates – I got into a terrible mess trying to generate a lot of AccountNames using this activity. I think it was made even worse by SQL doing an extra uniqueness lookup for that specific attribute.

**Run PowerShell Script Activity**

Powershell Script User: Worth mentioning that the FIMService service account is what is used to run the script by default.

Impersonation: there were problems with that and Craig’s PowerShell activity. The end result was a reg change was needed on each FIM Service server – if you didn’t make the reg change you could only impersonate members of the local Administrators group. The fact that a logon type has to be specified makes me think this might be the same. If so it would be worth expanding on the pre-reqs for the users that can be impersonated. It’s pretty common to want to say “any”, especially when it’s the original requestor we want to impersonate.

It would be really good to have a script template as a starting point, which shows how to send parameters to the script, and how to get information back from it.

I don’t agree with the comment about not using it in productions environments – in fact sometimes you have to (like when performing Exchange activities). Sounding a note of caution about spawning lots of sessions is worthwhile however.

On the comment about things failing – I use a lot of Invoke-Expression so I can set up the command I’m going to run, log the exact command, then run it, and capture the results. Try…Catch followed by testing we got a result is very helpful too. Anything to stop the script bombing out but returning a valid status. (Another thing to include – how to return a Failed status to the workflow – I normally just Throw the error message I want the request to store – that’s with Craig’s activity.)

I don’t see anything about the changes that need to be made to be able to run the FIMAutomation cmdlets. It’s possible that the config file changes are made as part of the setup, but we also need a person object for the FIMService service account that can login to the Portal and has the appropriate rights.

Finally – something on error reporting from the scripts would be good. Eg write to log fie, write the Event Log, return RequestStatusDetail to request.

**Activity Advanced Features Iteration**

Supported Activities does not mention Delete Resource.

**Comment by NileshGhodekar** [comment]
Jef, did you misss the XPath Search filter wiki? The usage is more generic that Queries. Also there were a couple of post on yammer on this subject, the latest being by Paul.

**Comment by JefTek** [comment]
Nilesh,

I must have missed this :)

So I can wrap the Xpath in something lie this, and if it returns Null does it not execute the query?

IIF(IsPresent([//WorkflowData/LocationCode]),"/Location[LocationCode="+[//WorkflowData/LocationCode] +"]",Null())

In other words, if I don't have a [//WorkflowData/LocationCode] value, don't issue the query?    So the Query and Activity wouldn't fire if  IsPresent([//WorkflowData/LocationCode]) returns false.

**Comment by NileshGhodekar** [comment]
#### Status - Build And Deployment Wiki
I've made additional clarifications around the IsAuthZ typo. We have also made some additional clarification on UpdateWorkflowXoml scripts after grate debate last night :). This item is considered closed.

**Comment by NileshGhodekar** [comment]
#### Status - Delay Activity Wiki

I've made some additional clarification in the implementation guidance on the FIM WF host behavior and added MSDN references. This item is can be opened only after Eugene does some testing :) :)

**Comment by NileshGhodekar** [comment]
#### Status - Create Resource Wiki

I've made some clarifications around the recommendation. I think this closed it, but you may defer on the recommendation or want to add more.

**Comment by NileshGhodekar** [comment]
#### Status - Generate Unique Value Wiki

I've moved Eugene's note inline and also added some addition clarifications. This items is considered closed.

**Comment by NileshGhodekar** [comment]
Jef, re:XPath expression, your syntax is a little wrong and the correct syntax will result it a "/" if the Null() is returned, which will then be an invalid XPath by FIM grammar. But WAL can always treat it as a null query. This feature may be for a future release :)

**Comment by NileshGhodekar** [comment]
##### Status - Run PowerShell Script Wiki

Updated wiki to address comments and much more.

@cwapshere 
re: Impersonation, I added a reference to the PS Connector for the permissions required. Also clarified that who gets impersonated and that there is no way to impersonate original requestor when running scripts.

re: script template, For the script input and output, I've added additional examples in the examples sections. The script `RunPSLoggingSample.ps1` that gets copied to the SolutionOutput folder post build is pretty much the starting template that everyone should be using.

re: production recommendation, unfortunately it stays. But I've elaborated the reasoning in much more detail so one can take an informed decision.

re: error handling and logging, I've elaborated that try/catch/trap may not help with avoiding errors and a pessimistic approach to script development is better suited to avoid aborted workflow.  The script error reporting / tracing was already covered but I've now moved the paragraph before the InvokeImmediateTermination script example for out-of-process scripts as I see that you missed it. The script InvokeImmediateTermination is already quite well instrumented even though it's hard to notice. Plug it in the WAL activity and if you don't believe :)

re: FIMAutomation cmdlets, I've added a liner in the implementation guidance that if WAL's native activities can't do it, report the issue. We were not able to find a use case for it when WAL is used.

I'm marking all comments on PSH activity closed now.

**Comment by EugeneSergeev** [comment]
FIMAutomation could be loaded in WAL PS Activity session. requires FIM Service config to be changed from having an external name to be http://localhost:5725

**Comment by NileshGhodekar** [comment]
I think Carol is already aware of this since she mentioned it also needs a Person object for FIM Service.

**Comment by EugeneSergeev** [comment]
correct

**Comment by cwapshere** [comment]
I agree the WAL takes away a lot of the need to work around the deficiencies of the OOB workflows with powershell - but I still think there are likely to be cases where you might want to run a script and write something back. In a WAL environment (earlier version) I have a "ScriptRunner" object type which calls out scripts with parameters set by workflow and then writes status and logging details back. I use it for running scripts that do Exchange and home folder creates, moves and archives. So a note about the changes that need to be made if you want to run FIMAutomation commands would just complete the picture.

**Comment by weyto** [comment]
Hi all.
I have tested the 3rd example (Delete All Groups Owned by User) on the Delete Resource Activity and it doesn’t work, when the Target is deleted. But if the Target Object is still available the target search filter is ok and the activity works.
When the Target object isn’t available in the FIM Service database (because the activity is triggered on an object deletion) then it is necessary to modify the target search filter to /Group[BackLink = ‘[//Request/Target]’]
Perhaps it is also useful to mention that the attribute BackLink must be of type string.

**Comment by JefTek** [comment]
Carol,  I'm going to take the topic of FIMAautomation and create a topic about this in the MIMWAL Technet forum.   I think it's something that could use some discussion as a pattern, and I think would be better served there.

**Comment by JefTek** [comment]
Thomas,  Thanks for the feedback!   We'll take a look at the example and see where we can improve it.

**Comment by NileshGhodekar** [comment]
@weyto, thanks for the feedback. re: DeleteResources activity, the workflow should be AuthZ or Action depending on the case, I've added a note to these examples. I've also changed BackLink to Owner (copy paste error basically) so that you know it must be a Reference :)

**Comment by NileshGhodekar** [comment]
@cwapshere , re: FIMAutomation, the design pattern we are promoting is that you do only the bare minimum non-FIM work in your script and comeback into WAL to do the rest of the work. We have not found a usecase to do otherwise in FIMWAL 2, but if there is we'll definitely investigate the possibility of adding any missing feature / function.

**Comment by NileshGhodekar** [comment]
@cwapshere, btw, I've just uploaded the Troubleshooting wiki and it has examples of errors / debug messages from the first PowerShell script listed on the PowerShell activity wiki.

**Comment by weyto** [comment]
@NileshGhodekar, thanks for the answer. Nice solution to trigger the deletion in AuthZ. Make it sence to modify the Register.ps1 in the github version? Per default the Delete Activity isn't a AuthZ Activity.

**Comment by EugeneSergeev** [comment]
I would strongly recommend to find another solution for 2 reasons:
1. AuthZ is not applied to FIM Sync requests
2. there could be another AuthZ that will deny request seconds later

instead, after the owner will deleted trigger another policy to fire a WF for all groups w/o owners.

**Comment by NileshGhodekar** [comment]
That is correct. Hence I never thought the initial verbiage will be interpreted as when the object is gone from the system. You'd typically delete a user at a future date (disable and delete) and action workflow is the most common use case. For the same reason the default for DeleteResources activity is Action workflow only (which you can always change).

**Comment by Amethi** [comment]
Guys, I'm so glad to see this open-sourced, nice one!

One thing on documentation - it might be worth pointing out that the **Word()** function in MIMWAL works differently to the native Word function built into FIM/MIM. This caught me out when working with the framework in it's earlier incarnation.

The difference is:
  - FIM/MIM Word() uses a 1 based index
  - MIMWAL Word() uses a 0 based index

Reference:
  - [Functions Reference for FIM 2010](https://technet.microsoft.com/en-us/library/ff800820(v=ws.10).aspx)
  - [MIMWAL Word() Source](https://github.com/Microsoft/MIMWAL/blob/master/src/WorkflowActivityLibrary/Common/ExpressionFunction.cs#L3679)

**Comment by NileshGhodekar** [comment]
Nice to see you Jay! Yes, this is true of all the string functions. There is a special mention of this in [CONTRIBUTING.md](https://github.com/Microsoft/MIMWAL/blob/master/CONTRIBUTING.md), but yes this needs to be called out in the functions wiki as well.

**Comment by iecompattest** [comment]
good

**Comment by Amethi** [comment]
Hey Nilesh :)

**Comment by NileshGhodekar** [comment]
Guys, I'm closing this issue. The only issue opened was better examples for some of the functions. If someone shares them, we'll update the wiki. Since there was a good amount of discussion on RunPowerShellScript activity in this Issue log, I also want to mention that in the latest check-in there are some changes made to how RunPowerShellScript activity handles errors and the Wiki is updated as below:

"Prior to build 2.15.305.0, the RunPowerShellScript activity treated any error in the script as fatal errors and aborted the workflow. As of build 2.15.305.0, the RunPowerShellScript activity aborts the workflow execution only when an unhandled or explicit exception is thrown from the PowerShell script. All other errors are simply logged in the event log. However, depending on the logic encoded in the PowerShell script, a non-terminating error may be equally detrimental to your business objective. It is now your responsibility to check for any errors (the `$Error` variable) and `throw` an exception for errors that are non-terminating but logically fatal to abort the workflow."
